cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServer.java

```diff
@@ -0,0 +1,191 @@
+                       @Named(Constants.AppFabric.SERVER_ADDRESS) InetAddress hostname,
+                       @Named(Constants.Preview.HANDLERS_BINDING) Set<HttpHandler> handlers,
+                       @Nullable MetricsCollectionService metricsCollectionService,
+                       ProgramRuntimeService programRuntimeService,
+                       ApplicationLifecycleService applicationLifecycleService,
+                       ProgramLifecycleService programLifecycleService,
+                       SystemArtifactLoader systemArtifactLoader) {
+    this.hostname = hostname;
+    this.discoveryService = discoveryService;
+    this.configuration = configuration;
+    this.handlers = handlers;
+    this.metricsCollectionService = metricsCollectionService;
+    this.programRuntimeService = programRuntimeService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+    this.systemArtifactLoader = systemArtifactLoader;
+  }
+
+  /**
+   * Configures the Preview pre-start.
+   */
+  @Override
+  protected void startUp() throws Exception {
+    LoggingContextAccessor.setLoggingContext(new ServiceLoggingContext(Id.Namespace.SYSTEM.getId(),
+                                                                       Constants.Logging.COMPONENT_NAME,
+                                                                       Constants.Service.PREVIEW_HTTP));
+    Futures.allAsList(
+      applicationLifecycleService.start(),
+      systemArtifactLoader.start(),
+      programRuntimeService.start(),
+      programLifecycleService.start()
+    ).get();
+
+    // Create handler hooks
+    ImmutableList.Builder<HandlerHook> builder = ImmutableList.builder();
+    builder.add(new MetricsReporterHook(metricsCollectionService, Constants.Service.PREVIEW_HTTP));
+
+
+    // Run http service on random port
+    httpService = new CommonNettyHttpServiceBuilder(configuration)
+      .setHost(hostname.getCanonicalHostName())
+      .setHandlerHooks(builder.build())
+      .addHttpHandlers(handlers)
+      .setConnectionBacklog(configuration.getInt(Constants.AppFabric.BACKLOG_CONNECTIONS,
+                                                 Constants.AppFabric.DEFAULT_BACKLOG))
+      .setExecThreadPoolSize(configuration.getInt(Constants.AppFabric.EXEC_THREADS,
+                                                  Constants.AppFabric.DEFAULT_EXEC_THREADS))
+      .setBossThreadPoolSize(configuration.getInt(Constants.AppFabric.BOSS_THREADS,
+                                                  Constants.AppFabric.DEFAULT_BOSS_THREADS))
+      .setWorkerThreadPoolSize(configuration.getInt(Constants.AppFabric.WORKER_THREADS,
+                                                    Constants.AppFabric.DEFAULT_WORKER_THREADS))
+      .build();
+
+    // Add a listener so that when the service started, register with service discovery.
```

Comment:
Why need to be so complicated? Since this class owns the http service, it can simply start the http server and block on it, then announce it to the discovery service. Similarly for stop as well (cancel service discovery followed by stopping server).